### PR TITLE
Calling Super in Swift

### PR DESCRIPTION
**Pull request by steipete**

### CODE REVIEW

File: _posts/2020-06-10-calling-super-at-runtime.md [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+     };
+     objc_msgSendSuper2(&_super, _cmd);
+ }
+ ```
+
+In compiled code, there’s a lookup table so that `object_getClass` doesn’t need to be called, but you see the principle. A struct is created, and `objc_msgSendSuper2` is called with it. The method is automatically dynamic, since UIKit is written in Objective-C, so the Swift compiler knows it needs to use dynamic dispatch. 
+
+## objc_msgSendSuper vs. objc_msgSendSuper2
+
+There are two versions of the super call, and the difference is minor but important: `objc_msgSendSuper` starts looking at the current class, which would cause an endless loop in above code, while `objc_msgSendSuper2` looks for the superclass.
```
Comment (deutschbitte) [inline _posts/2020-06-10-calling-super-at-runtime.md:41]: Super in backticks or no?

File: _posts/2020-06-10-calling-super-at-runtime.md [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+        };
+        return ((id(*)(struct objc_super *, SEL, va_list))objc_msgSendSuper2)(&super, selector, argp);
+    }), types);
+```
+
+This works, and [we’ve been shipping code](https://pspdfkit.com/blog/2019/swizzling-in-swift/) like this for a while. For InterposeKit, I wanted to write the same in Swift.
+
+## Super in Swift
+
+An additional goal was to be “pure” Swift, not because I’m a purist, but because SwiftPM doesn’t yet support mixed language projects.[^4] We can’t write a C header to import `objc_msgSendSuper2`, but we sure can look it up at runtime:
```
Comment (deutschbitte) [inline _posts/2020-06-10-calling-super-at-runtime.md:67]: Be writing? Be using?

"An additional goal was to be “pure” Swift,"

File: _posts/2020-06-10-calling-super-at-runtime.md [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+#endif
+```
+
+Previously, it was declared as a function that took `id`, `SEL`, and variadic arguments, returning `id` — now it takes and returns void. Why the change? The short version is that there is no guarantee that the ABI for variadic function matches the ABI for a function with a mixed number of arguments. If you just pass pointers, this did match reasonably enough to mostly “just work.”
+
+However, the ARM64 ABI [is more complicated.](https://blog.nelhage.com/2010/10/amd64-and-va_arg/), and variadic arguments are passed on the stack. Without casting, even a trivial use of `objc_msgSend` will result in a crash. There is an interesting article by Mike Ash entitled [objc_msgSend’s New Prototype](https://www.mikeash.com/pyblog/objc_msgsends-new-prototype.html). Mike’s blog is brilliant, and I’m extremely happy that he still writes new posts from time to time, despite now working at Apple.
+
+## Accepting Assembly
+
+Usually the compiler takes care of casting `objc_msgSendSuper` for us — this isn’t something that it can do when we try to do this at runtime. The only way to call this in a correct without getting lucky is if we write the call in assembly.
```
Comment (deutschbitte) [inline _posts/2020-06-10-calling-super-at-runtime.md:125]: By "in a correct" are you missing a word? Or do you want to say "call this correctly?"

BTW, I got rid of blind luck even with quotes because it's ableist language. I think the first instance where you're directly quoting the tweet is _OK_ but the second time I'd rephrase it. 

"The only way to call this in a correct without getting lucky is if we write the call in assembly."

File: _posts/2020-06-10-calling-super-at-runtime.md [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+First of all, [assembly is hard](https://twitter.com/steipete/status/1270035179424399360?s=21), but it’s a useful skill that will make you better at debugging, so I’ve approached this as a “fun“ challenge. The most important part to know is what each register does. To keep things simple, we focus on ARM64 in this article.
```
Comment (deutschbitte) [inline _posts/2020-06-10-calling-super-at-runtime.md:127]: You've approached what as a fun project? Assembly in general or this specific topic in the blog post?

"First of all, [assembly is hard](https://twitter.com/steipete/status/1270035179424399360?s=21), but it's a useful skill that will make you better at debugging, so I've approached this as a "fun" challenge."

File: _posts/2020-06-10-calling-super-at-runtime.md [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+void trampolineFunction(void) { \
+asm volatile ("add " #selfLocation ", " #selfLocation ", #" #offset "\n\t" \
+"b " #msgSendSuperFunction "\n\t" \
+: : : "x0", "x1"); \
+}
+```
+
+## Class Variable Layout
+
+What this code here so cleverly does is that it simply adds eight bytes to the location of self. The layout of classes looks like this:
```
Comment (deutschbitte) [inline _posts/2020-06-10-calling-super-at-runtime.md:162]: Self in backticks?

"What this code here so cleverly does is that it simply adds eight bytes to the location of self."

File: _posts/2020-06-10-calling-super-at-runtime.md [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+               ├───────────────────┤
+               │Trampoline Data #3 │ 0x6000 Read, Write
+               └───────────────────┘               
+```
+
+With a fixed offset, we can reach the corresponding data from entry, and we can read variables as needed. This is how `imp_implementationWithBlock` works, and luckily it’s also [open source](https://github.com/0xxd0/objc4/blob/master/objc4/runtime/objc-block-trampolines.mm) — but there be [dragons](https://twitter.com/steipete/status/1269912889097363457?s=21). Landon Fuller [reimplemented this](https://landonf.org/code/objc/imp_implementationWithBlock.20110413.html) back when it was introduced in iOS 4.3 and explains the principles really well.
+
+## Tail Calling
+
+There’s a lot of logic required to correctly manage tables, and things need locking to make it thread-safe. I decided that this is gonna be the backup plan and tried a more direct approach to just fetch everything at runtime.
```
Comment (deutschbitte) [inline _posts/2020-06-10-calling-super-at-runtime.md:205]: "It" being the code?

"There's a lot of logic required to correctly manage tables, and things need locking to make it thread-safe."

File: _posts/2020-06-10-calling-super-at-runtime.md [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+The principle: We save the registers that we might spill, fill the struct at runtime, restore the registers, and then perform the tail call. This sounds simple now that I write it up, but it caused serious headaches at first. 
+
+Specifically, I tried to use the stack to generate the struct, which breaks stack-based parameter passing. I tried calling malloc in asm, but since that requires free, I couldn’t do the tail-call optimization anymore. And I encountered oh so many crashes because I didn’t really understand what it means to align the stack pointer on 16 bits.
```
Comment (deutschbitte) [inline _posts/2020-06-10-calling-super-at-runtime.md:209]: What is "requires free" saying? And malloc in backticks?

"I tried calling malloc in asm, but since that requires free, I couldn't do the tail-call optimization anymore."

File: _posts/2020-06-10-calling-super-at-runtime.md [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+```
+// push {x0-x8, lr} (call params are: x0-x7)
+"stp x8, lr, [sp, #-16]!\n" // lr = link register
+"stp x6, x7, [sp, #-16]!\n"
+"stp x4, x5, [sp, #-16]!\n"
+"stp x2, x3, [sp, #-16]!\n" // push x3, then x2
+"stp x0, x1, [sp, #-16]!\n" // push x1, then x0
+```
+
+`stp` saves a pair of registers (2*8=16 byte) on the stack, and it also automatically decrements the stack pointer. The stack on most architectures grows downward from max to 0, so via decrementing, we reserve memory:
```
Comment (deutschbitte) [inline _posts/2020-06-10-calling-super-at-runtime.md:222]: Zero in backticks?

"The stack on most architectures grows downward from max to 0,"

File: _posts/2020-06-10-calling-super-at-runtime.md [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+
+```
+// fetch filled struct objc_super, call with self + _cmd
+"bl _ITKReturnThreadSuper \n"
+```
+
+`bl` means “branch with link.” and it calls a function, in this case a C function. The same call arguments exist here, so the first parameter will be `self`, the second will be `_cmd`. `bl` and `b` are [similar](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0068b/CIHFDDAF.html), however `bl` stores the address of the next instruction into the `lr` register, so the called function can jump back via `ret`.
```
Comment (deutschbitte) [inline _posts/2020-06-10-calling-super-at-runtime.md:229]: By "so" do you mean it does that so that the function can jump back, or does the function jump back as a result of this happening?

"however `bl` stores the address of the next instruction into the `lr` register, so the called function can jump back via `ret`."

File: _posts/2020-06-10-calling-super-at-runtime.md [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+
+Also see that `object_getClass`, and not the class method, is used here. While the latter can be overridden so a class “lies” about its type, this always returns the correct type. This is important since Apple uses this trick for key-value coding, which creates a subclass at runtime but also overrides `class` to hide this fact:
+
+```
+// first param is now struct objc_super (x0)
+// protect returned new value when we restore the pairs
+"mov x9, x0\n"
+```
+
+Once we return, in ARM64, the return value is in `x0`. We temporarily store this in the “scratch space” register set; `x9`-`x15` are free to use. Another word for this is caller-saved or clobbered. Why do we do that when `x0` is already exactly what we want? Because on `ARM64`, we can only operate on the stack in 16 bytes, so we always restore pairs of registers:
```
Comment (deutschbitte) [inline _posts/2020-06-10-calling-super-at-runtime.md:257]: Should ARM64 be in backticks or not?

"Once we return, in ARM64, the return value is in `x0`. We temporarily store this in the “scratch space” register set; `x9`-`x15` are free to use. Another word for this is caller-saved or clobbered. Why do we do that when `x0` is already exactly what we want? Because on `ARM64`, we can only operate on the stack in 16 bytes, so we always restore pairs of registers:"

File: _posts/2020-06-10-calling-super-at-runtime.md [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+```
+// pop {x0-x8, lr}
+"ldp x0, x1, [sp], #16\n"
+"ldp x2, x3, [sp], #16\n"
+"ldp x4, x5, [sp], #16\n"
+"ldp x6, x7, [sp], #16\n"
+"ldp x8, lr, [sp], #16\n"
+```
+
+While there are ways around this, they are less elegant and require even more assembly. We are now back at the state of before doing anything, almost ready for the super call:
```
Comment (deutschbitte) [inline _posts/2020-06-10-calling-super-at-runtime.md:268]: Meaning that almost ready for the super call is the state of before doing anything, or these are two things 1) we're back at the state 2) we're almost ready for the super call?

"We are now back at the state of before doing anything, almost ready for the super call."

File: _posts/2020-06-10-calling-super-at-runtime.md (hunk 1 above)
Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:41]: I'd say so yes

File: _posts/2020-06-10-calling-super-at-runtime.md (hunk 2 above)
Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:67]: An additional goal [for this library] was to be [written in] "pure" Swift

Not sure this is needed for context, it's a bit of an inside joke.

File: _posts/2020-06-10-calling-super-at-runtime.md (hunk 3 above)
Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:125]: in a correct way

Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:125]: Good point! I wanted to be funny here.

File: _posts/2020-06-10-calling-super-at-runtime.md (hunk 4 above)
Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:127]: The "How can I call super at runtime" was a fun project/challenge.

File: _posts/2020-06-10-calling-super-at-runtime.md (hunk 5 above)
Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:162]: yes

File: _posts/2020-06-10-calling-super-at-runtime.md (hunk 6 above)
Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:205]: it = [the trampoline management code]

File: _posts/2020-06-10-calling-super-at-runtime.md (hunk 7 above)
Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:209]: I tried calling `malloc` in asm, but since that requires calling `free`

File: _posts/2020-06-10-calling-super-at-runtime.md (hunk 8 above)
Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:222]: In that case it's a memory address, so it's a normal number, wouldn't use backticks for numbers in general.

File: _posts/2020-06-10-calling-super-at-runtime.md (hunk 9 above)
Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:229]: Better?: ... register, therefore the called function can jump back via `ret`.

Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:229]: it can jump back because bl stores the address where to jump back (b doesn't)

File: _posts/2020-06-10-calling-super-at-runtime.md (hunk 10 above)
Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:257]: It's the architecture name - no backticks.

File: _posts/2020-06-10-calling-super-at-runtime.md (hunk 11 above)
Comment (steipete) [reply _posts/2020-06-10-calling-super-at-runtime.md:268]: After restoring the registers, we're now almost ready for the super call:
